File: testCode/js/Mouse.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
// import { Vector2 } from "three";
import Common_1 from "./Common.js";

var Vector2 = THREE.Vector2
var Mouse = /** @class */ (function () {
    function Mouse() {
        this.mouseMoved = false;
        this.coords = new Vector2();
        this.coords_old = new Vector2();
        this.diff = new Vector2();
        this.timer = null;
        this.count = 0;
    }
    Mouse.prototype.init = function () {
        document.body.addEventListener("mousemove", this.onDocumentMouseMove.bind(this), false);
        document.body.addEventListener("touchstart", this.onDocumentTouchStart.bind(this), false);
        document.body.addEventListener("touchmove", this.onDocumentTouchMove.bind(this), false);
    };
    Mouse.prototype.setCoords = function (x, y) {
        var _this = this;
        if (this.timer)
            clearTimeout(this.timer);
        this.coords.set((x / Common_1.width) * 2 - 1, -(y / Common_1.height) * 2 + 1);
        this.mouseMoved = true;
        this.timer = setTimeout(function () {
            _this.mouseMoved = false;
        }, 100);
    };
    Mouse.prototype.onDocumentMouseMove = function (event) {
        this.setCoords(event.clientX, event.clientY);
    };
    Mouse.prototype.onDocumentTouchStart = function (event) {
        if (event.touches.length === 1) {
            // event.preventDefault();
            this.setCoords(event.touches[0].pageX, event.touches[0].pageY);
        }
    };
    Mouse.prototype.onDocumentTouchMove = function (event) {
        if (event.touches.length === 1) {
            // event.preventDefault();
            this.setCoords(event.touches[0].pageX, event.touches[0].pageY);
        }
    };
    Mouse.prototype.update = function () {
        this.diff.subVectors(this.coords, this.coords_old);
        this.coords_old.copy(this.coords);
        if (this.coords_old.x === 0 && this.coords_old.y === 0)
            this.diff.set(0, 0);
    };
    return Mouse;
}());
const _default = new Mouse();
export { _default as default };

File: lua/5.3/src/Mouse.lua
local window = js.global
local document = window.document
local THREE = window.THREE

local Common = require("src.Common")
local new = require("utils.new")

return (function()
    local self = {}

    self.mouseMoved = false
    self.coords = new(THREE.Vector2)
    self.coords_old = new(THREE.Vector2)
    self.diff = new(THREE.Vector2)
    self.timer = nil
    self.count = 0

    self.mouseDownX = 100
    self.mouseDownY = 100
    self.mouseUpX = 200
    self.mouseUpY = 200

    self.mouseDownBind = false
    self.mouseUpBind = false

    self.init = function()
        document.body:addEventListener('mousemove', self.onDocumentMouseMove, false)
        document.body:addEventListener('mousedown', self.onDocumentMouseDown, false)
        document.body:addEventListener('mouseup', self.onDocumentMouseUp, false)
        document.body:addEventListener('touchstart', self.onDocumentTouchStart, false)
        document.body:addEventListener('touchmove', self.onDocumentTouchMove, false)
    end

    self.setCoords = function(x, y)
        if (self.timer) then window:clearTimeout(self.timer) end
        self.coords:set((x/Common.width) * 2 - 1, -(y/Common.height) * 2 + 1)
        self.mouseMoved = true
        self.timer = window:setTimeout(function()
            self.mouseMoved = false
        end, 100)
    end

    self.onDocumentMouseMove = function(_, event)
        self.setCoords(event.clientX, event.clientY)
    end

    self.onDocumentMouseDown = function(_, event)
        self.mouseDown = true
        self.mouseDownX = event.clientX
        self.mouseDownY = event.clientY

        if (self.mouseDownBind) then self.mouseDownBind() end
    end

    self.onDocumentMouseUp = function(_, event)
        self.mouseDown = false
        self.mouseUpX = event.clientX
        self.mouseUpY = event.clientY

        if (self.mouseUpBind) then self.mouseUpBind() end
    end

    self.onDocumentTouchStart = function(_, event)
        if (event.touches.length == 1) then
            self.setCoords(event.touches[0].pageX, event.touches[0].pageY)
        end
    end

    self.onDocumentTouchMove = function(_, event)
        if (event.touches.length == 1) then
            self.setCoords(event.touches[0].pageX, event.touches[0].pageY)
        end
    end

    self.update = function()
        self.diff:subVectors(self.coords, self.coords_old)
        self.coords_old:copy(self.coords)

        if (self.coords_old.x == 0 and self.coords_old.y == 0) then self.diff:set(0, 0) end
    end

    return self
end)()